cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperScaler.java

```diff
@@ -0,0 +1,351 @@
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Class for scaling Zookeeper 3.5 using the ZookeeperAdmin client
+ */
+public class ZookeeperScaler implements AutoCloseable {
+    private static final Logger log = LogManager.getLogger(ZookeeperScaler.class);
+
+    private final Vertx vertx;
+    private final ZooKeeperAdminProvider zooAdminProvider;
+
+    private final String zookeeperConnectionString;
+
+    private final long operationTimeoutMs;
+
+    private final String trustStorePassword;
+    private final File trustStoreFile;
+
+    private final String keyStorePassword;
+    private final File keyStoreFile;
+
+    private ZooKeeperAdmin zkAdmin;
+
+    /**
+     * ZookeeperScaler constructor
+     *
+     * @param vertx                         Vertx instance
+     * @param zookeeperConnectionString     Connection string to connect to the right Zookeeper
+     * @param clusterCaCertSecret           Secret with Kafka cluster CA public key
+     * @param coKeySecret                   Secret with Cluster Operator public and private key
+     * @param operationTimeoutMs            Operation timeout
+     *
+     * @return  ZookeeperScaler instance
+     */
+    protected ZookeeperScaler(Vertx vertx, ZooKeeperAdminProvider zooAdminProvider, String zookeeperConnectionString, Secret clusterCaCertSecret, Secret coKeySecret, long operationTimeoutMs) {
+        log.debug("Creating Zookeeper Scaler for cluster {}", zookeeperConnectionString);
+
+        this.vertx = vertx;
+        this.zooAdminProvider = zooAdminProvider;
+        this.zookeeperConnectionString = zookeeperConnectionString;
+        this.operationTimeoutMs = operationTimeoutMs;
+        trustStorePassword = new String(Util.decodeFromSecret(clusterCaCertSecret, Ca.CA_STORE_PASSWORD), StandardCharsets.US_ASCII);
+        trustStoreFile = Util.createFileStore(getClass().getName(), "p12", Util.decodeFromSecret(clusterCaCertSecret, Ca.CA_STORE));
+        keyStorePassword = new String(Util.decodeFromSecret(coKeySecret, "cluster-operator.password"), StandardCharsets.US_ASCII);
+        keyStoreFile = Util.createFileStore(getClass().getName(), "p12", Util.decodeFromSecret(coKeySecret, "cluster-operator.p12"));
+    }
+
+    /**
+     * Scales Zookeeper to defined number of instances.
+     * It generates new configuration according to the desired number of nodes and updates Zookeeper configuration.
+     *
+     * @param scaleTo   Number of Zookeeper nodes which should be used by the cluster
+     *
+     * @return          Future which succeeds / fails when the scaling is finished
+     */
+    public Future<Void> scale(int scaleTo) {
+        return connect()
+                .compose(ignore -> getConfig())
+                .compose(servers -> scaleTo(servers, scaleTo))
+                .compose(ignore -> closeConnectionAsync());
```

Comment:
The connection won't get closed in if there is a failed `Future`.

Reply:
It is closed from KafkaAssemblyOperator in that case. But yeah, it is probably better to handle it already here.

Reply:
I added assitional handling here as well.